File: javascript/scaffold.js
module('users.timfelgentreff.babelsberg.testsuite').
requires('lively.TestFramework',
         'users.timfelgentreff.babelsberg.constraintinterpreter',
         'users.timfelgentreff.z3.CommandLineZ3').
toRun(function() {

TestCase.subclass('users.timfelgentreff.babelsberg.testsuite.SemanticsTests', {
    fieldEquals: function(o1, o2) {
        if (!o1) return false;
        if (!o2) return false;
        for (var key in o1) {
            if (key[0] !== "$" && key[0] !== "_") {
                if (typeof(o1[key]) == "object") {
                    this.fieldEquals(o2[key], o1[key]);
                } else {
                    if (o1[key] !== o2[key]) return false;
                }
            }
        }
        for (var key in o2) {
            if (typeof(o2[key]) == "object") {
                this.fieldEquals(o2[key], o1[key]);
            } else {
                if (o1[key] !== o2[key]) return false;
            }
        }
        return true;
    },
    one: function(self) {
        return 1.0
    },
    double: function(self) {
        return 2 * self;
    },
    Require_min_balance: function(self, acct, min) {
        always: { acct.balance > min }
    },
    Has_min_balance: function(self, acct, min) {
        return acct.balance > min;
    },
    Point: function(self, x, y) {
        return {x: x, y: y}
    },
    center: function(self) {
        return this.divPtScalar(this.addPt(self.upper_left, self.lower_right), 2);
    },
    addPt: function(self, other) {
        return this.Point(null, self.x + other.x, self.y + other.y);
    },
    divPtScalar: function(self, scale) {
        return this.Point(null, self.x / scale, self.y / scale);
    },
    ptEq: function(self, other) {
        return self.x == other.x && self.y == other.y
    },
    Test: function(self, i) {
        ctx = {i: i};
        bbb.always({
            priority: "medium",
            ctx: {
                ctx: ctx,
                _$_self: this.doitContext || this
            }
        }, function() {
            return ctx.i == 5;;
        });
        return ctx.i + 1
    },
    MutablePointNew: function(self, x, y) {
        return new Object({x: x, y: y});
    },
    WindowNew: function(self) {
        return new Object({window: true});
    },
    CircleNew: function(self) {
        return new Object({circle: true});
    },
    // TODO: others


INSERTHERE

});
});

File: ruby/scaffold.rb
require "libz3"

class Helper < ConstraintObject
  def hRecord(hash)
    o = Object.new
    hash.each_pair do |k,v|
      o.instance_variable_set(:"@#{k}", v)
      o.singleton_class.send(:attr_accessor, k)
    end
    o.define_singleton_method "inspect" do
      hash.keys.inject("{") do |acc, i|
        "#{acc} #{i}: #{o.instance_variable_get('@' + i.to_s)}"
      end + "}"
    end
    o.define_singleton_method "values" do
      hash.keys.inject("{") do |acc, i|
        "#{acc} #{i}: #{o.instance_variable_get('@' + i.to_s)}"
      end + "}"
    end
    o.singleton_class.send("alias_method", :to_s, :inspect)

    o.define_singleton_method "==" do |o|
      if o.kind_of? Struct
        return hash.keys.all? do |var|
          self.instance_variable_get("@#{var}") == o.instance_variable_get("@#{var}")
        end
      else
        return self.equal? o
      end
    end
    return o
  end

  IRecordTypes = {}
  def iRecord(hash)
    attrs = hash.keys
    name = "IRecord_#{attrs.join}"
    klass = IRecordTypes[name]
    unless klass
      IRecordTypes[name] = klass = Struct.make_struct(name, attrs)
      attrs.each { |a| klass.send("attr_accessor", a) }

      klass.send(:define_method, "inspect") do
        attrs.inject("{") do |acc, i|
          "#{acc} #{i}: #{self.instance_variable_get('@' + i.to_s)}"
        end + "}"
      end
      klass.send("alias_method", :to_s, :inspect)
    end
    o = klass.new
    hash.each_pair { |k,v| o[k] = v }
    o
  end
end

$last_exception = nil
def assert(value, var, expected, got)
  raise $last_exception ? $last_exception : "Expected #{var} to equal #{expected}, got #{got}" unless value
end
def assert_identical(value, e1, e2)
  raise $last_exception ? $last_exception : "Expected #{e1} to be identical to #{e2}" unless value
end

class Z3::Z3Pointer
  def double
    self * 2
  end
end

class NilClass
  def MutablePointNew(x,y)
    Helper.new.hRecord(x: x, y: y)
  end

  def Point(x,y)
    Helper.new.iRecord(x: x, y: y)
  end

  def WindowNew
    Helper.new.hRecord(window: true)
  end

  def CircleNew
    Helper.new.hRecord(circle: true)
  end

  def Testpointxequals5(p)
    always { p.x == 5 }
    return p
  end

  def Test(i)
    always(priority: :medium) { i == 5 }
    return i + 1
  end

  def Has_min_balance(acct, min)
    return acct.balance > min
  end

  def Require_min_balance(acct, min)
    always { acct.balance > min }
  end
end

class Object
  def center
    self.upper_left.addPt(self.lower_right).divPtScalar(2)
  end

  def addPt(pt)
    Helper.new.iRecord(x: self.x + pt.x, y: self.y + pt.y)
  end

  def divPtScalar(scale)
    Helper.new.iRecord(x: self.x / scale, y: self.y / scale)
  end

  def ptEq(pt)
    pt.x == self.x && pt.y == self.y
  end
end


INSERTHERE


errors = []
methods.
  select { |m| m.to_s.start_with? "test" }.
  sort_by { |e| e.to_s.sub("test", "").to_i }.
  each do |m|
  begin
    send(m)
  rescue Exception => e
    errors << e.to_s + ":\n      " + e.backtrace[1..-6].join("\n      ")
  end
end

print "#{errors.size} Errors:\n#{errors.join("\n  ")}\n"
